Intelligun - https://baystation12.net/forums/threads/prototype-weapon…

Esoudiere opened the pull request:
-Intelligun
Kind of forgot to add it in earlier, woopsies.
  -Assembly System - https://baystation12.net/forums/threads/suggestion-assembly-remake.977/
Logic circuits & modular 'upgrading' will hopefully be done eventually.
 -Both updated from ~5 month old codebase, I tested them briefly, seemed to work. Might still need a bit of updating though.
-Kinda new at coding in general, I imagine there being alot of optimisation errors that could be fixed.

Replaced grenade types with /device/assembly_holder/grenade, done in many files.
Edited TextFilters.dm, added RadioChat filter.
Edited disposal.dm & airlock.dm for use with assemblies
Added assembly.dm under wires for assembly wires.
Modified wire.dm to make some changes needed to make assembly.dm work available to others.
Modified glass.dm to include /assembly_holder in allowed containers

Esoudiere committed:
Intelligun - https://baystation12.net/forums/threads/prototype-weapon-for-hos.829/
  - Kind of forgot to add it in earlier, woopsies.
Assembly System - https://baystation12.net/forums/threads/suggestion-assembly-remake.977/
  - Logic circuits & modular 'upgrading' will hopefully be done eventually.
-Both updated from ~5 month old codebase, I tested them briefly, seemed to work. Might still need a bit of updating though.
-Kinda new at coding in general, I imagine there being alot of optimisation errors that could be fixed.

Mloc commented:
Lots of CI errors, https://travis-ci.org/Baystation12/Baystation12/builds/109863303
+8,848 −157 is very worrying, are you not deleting old code?
Can you list all the changes in the PR and commit message?

Inforsaken commented:
Sorry, maybe I should have split it into two parts to make it more clear.

Esoudiere committed:
Fixed some travis errors..

Esoudiere committed:
Fixed type errors in map

Esoudiere committed:
Assumed .dme wasn't needed. Apparently it is?

Esoudiere committed:
Renamed .dme (this is fun)

Mloc commented:
can you delete the files you're removing and not use a new/ dir prefix
and also not commit the map changes in the dme

Esoudiere committed:
Attempt to fix travis explosion

comma commented on code/modules/assembly/new/assembly.dm:
```diff
@@ -0,0 +1,333 @@
+	wires |= WIRE_ASSEMBLY_PASSWORD
+	active_wires = wires
+	wire_holder = new (src, (wire_num+1)) // Hacky way of doing it.
+	spawn(40) if(radio)
```
This should go in initialize(), like radios do that https://github.com/Baystation12/Baystation12/blob/dev/code/game/objects/items/devices/radio/radio.dm#L77
That proc is called after everything is created, so there's no need for hacky waits like this.

comma commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+	var/list/indexes_to_pick = list()
+	var/obj/item/device/assembly/A = holder
+	//Generate our indexes
+	for(var/i = 1; i < MAX_FLAG && indexes_to_pick.len < wire_count; i += i)
```
There's some bit touching macroes hidden in list helpers for stuff like that
https://github.com/Baystation12/Baystation12/blob/ed932901a620cef0ec861c81ad7805468662f693/code/_helpers/lists.dm#L357
```
#define BITTEST(bitfield,index)  ((bitfield)  &   (1 << (index)))
#define BITSET(bitfield,index)   (bitfield)  |=  (1 << (index))
#define BITRESET(bitfield,index) (bitfield)  &= ~(1 << (index))
#define BITFLIP(bitfield,index)  (bitfield)  ^=  (1 << (index))
```

comma commented on code/modules/assembly/new/wire_safety.dm:
```diff
@@ -0,0 +1,81 @@
+				if("Pulse Protection")
+					pulse = !pulse
+		if(href_list["index"])
+			if(!(protected_index.Remove(href_list["index"])))
```
What's this check for?
If you just need to have unique only values in protected index, you can do 
`protected_index |= href_list["index"]`

Esoudiere committed:
Trying yet again to find out the cause of travis errors
Removed /new dir
Replaced \blue \red etc with <span class>
Fixed a couple of things pointed out by Comma

Esoudiere committed:
Fixed indentation error

Ccomp5950 commented on baystation12.dme:
```diff
@@ -1933,9 +1981,4 @@
 #include "interface\interface.dm"
 #include "interface\skin.dmf"
 #include "maps\exodus-1.dmm"
-#include "maps\exodus-2.dmm"
```
No.

Esoudiere committed:
Added in missing z-levels.

Esoudiere committed:
Fixed broken html tags

Esoudiere committed:
Fixed runtime.

Esoudiere committed:
-Fixed to use attach_container proc

Esoudiere committed:
Type fix, forgot to compile to check.

Mloc commented:
can you actually tell us what this PR is doing in the main description and commit messages?

Techhead0 commented on baystation12.dme:
```diff
@@ -1140,8 +1187,6 @@
 #include "code\modules\events\wallrot.dm"
 #include "code\modules\examine\examine.dm"
-#include "code\modules\examine\stat_icons.dm"
-#include "code\modules\examine\descriptions\armor.dm"
```
These two files may have been removed from baystation12.dme in error.

Techhead0 commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+			else A.active_wires |= WIRE_ASSEMBLY_PROCESS
+		if(WIRE_ASSEMBLY_SAFETY)
+			if(A.active_wires & WIRE_ASSEMBLY_SAFETY) A.active_wires &= ~WIRE_ASSEMBLY_SAFETY
+			else A.active_wires |= WIRE_ASSEMBLY_SAFETY
```
Instead of having a bunch of near identical ifs for these, maybe consolidating them would be better. You can also use the `^` operator (Bitwise XOR) instead to simplify it.
```
		if(WIRE_POWER_RECEIVE, WIRE_POWER_SEND, WIRE_MISC_ACTIVATE, WIRE_MISC_SPECIAL, WIRE_ASSEMBLY_PROCESS, WIRE_ASSEMBLY_SAFETY)
			A.active_wires ^= index
```
I omitted the 'misc connection' and 'assembly password' wires from this, as it appears pulsing them manipulates other wires instead, but they should still use `^=`.

Techhead0 commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+				if(WIRE_ASSEMBLY_SAFETY)
+					L << "<span class='notice'>You [prob_text] \the [colour] wire is connected to \the Safety light.</span>"
+				if(WIRE_ASSEMBLY_PASSWORD)
+					L << "<span class='notice'>You [prob_text] \the [colour] wire is connected to \the Maintenance light.</span>"
```
Using `\the` before simple text probably isn't ideal.

Techhead0 commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -766,6 +767,10 @@ About the new airlock wires panel:
 		src.update_icon()
 	else if(istype(C, /obj/item/weapon/wirecutters))
+		if(attached_assembly)
+			usr << "<span class='notice'>You remove \the [attached_assembly]!</span>"
```
Use `user` instead of `usr`, when possible.

Techhead0 commented on code/modules/assembly/access_checker.dm:
```diff
@@ -0,0 +1,130 @@
+					access_levels_required = access_levels_requireds[index]
+					usr << "<span class='notice'>You set \the [src]'s required match count to \"[access_levels_required]\"</span>"
+
+		..()
```
This entire file uses relative pathing instead of absolute.

Techhead0 commented on code/modules/assembly/assembly.dm:
```diff
@@ -9,168 +9,323 @@
 	throw_speed = 3
 	throw_range = 10
-	origin_tech = list(TECH_MAGNET = 1)
+	origin_tech = "magnets=1"
```
`origin_tech` was changed from strings to a list. Don't change it back. (This applies elsewhere, as well.)

Techhead0 commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+		var/off = "<font color=#FF0000>off!</font>"
+		var/on = "<font color=#00FF00>on</font>"
+		var/blinking = "<font color=#FFA500>blinking!</font>"
+		if((A.wires & WIRE_DIRECT_RECEIVE) && (A.wires & WIRE_DIRECT_SEND))
```
`((x & a) && (x & b))` is equivalent to `(x & a & b)`.

Techhead0 commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+		if((A.wires & WIRE_DIRECT_RECEIVE) && (A.wires & WIRE_DIRECT_SEND))
+			if(A.active_wires & WIRE_DIRECT_RECEIVE && A.active_wires & WIRE_DIRECT_SEND)
+				. += "The <font color=#80000>dark red</font> light is [on]"
+			else if(!(A.active_wires & WIRE_DIRECT_RECEIVE) && !(A.active_wires & WIRE_DIRECT_SEND))
```
`(!(x & a) && !(x & b))` is equivalent to `!(x  & (a | b))`.

Techhead0 commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+			A.send_pulse_to_connected()
+		if(WIRE_RADIO_RECEIVE)
+			for(var/mob/M in view(1))
+				M.show_message("<small>*beep*</small>", 2)
```
Replacing these two lines with `audible_message("<small>*beep*</small>",,2)` will prevent deaf people from hearing the beeping.

Inforsaken replied:
Now it looks ridiculous; if(holder.holder) holder.holder.audible_message(blah blah). But fair point.

Techhead0 commented on code/datums/wires/assembly.dm:
```diff
@@ -0,0 +1,276 @@
+			else
+				. += "The <font color=#80000>dark red</font> light is [blinking]"
+		else if(A.wires & WIRE_DIRECT_RECEIVE || A.wires & WIRE_DIRECT_SEND)
+			. += "The <font color=#80000>dark red</font> light is [(A.active_wires & WIRE_DIRECT_RECEIVE || A.active_wires & WIRE_DIRECT_SEND)? "[on]" : "[off]"]"
```
`(x & a || x & b)` is equivalent to `(x & (a | b))`.

Techhead0 commented on code/modules/assembly/mousetrap.dm:
```diff
@@ -20,108 +23,114 @@
+				cable.use(10)
+				user.visible_message("<span class='warning'>\The [user] modifies \the [src]!</span>", "<span class='notice'>You add some wiring to \the [src], allowing it to trigger devices when tripped!</span>")
+				var/obj/item/device/assembly/mousetrap/pressure_plate/trap = new()
+				trap.loc = src.loc
```
Don't set `loc` directly, as it bypasses `Entered()` and `Exited()`. Use the `forceMove()` proc instead, eg. `trap.forceMove(src.loc)`.

Techhead0 commented on code/modules/assembly/assembly.dm:
```diff
@@ -9,168 +9,323 @@
+	if(istype(O, /obj/item/weapon/wirecutters))
+		if(!wires) return
+		user.set_machine(src)
+		wire_holder.Interact(user)
```
You don't need separate if statements here.
```
if(istype(O, /obj/item/device/multitool) || istype(O, /obj/item/weapon/wirecutters))
		if(!wires) return
		user.set_machine(src)
		wire_holder.Interact(user)
```

Techhead0 commented on code/modules/assembly/holder.dm:
```diff
@@ -1,277 +1,850 @@
+						removing = null
+					else
+						user.visible_message("<span class='notice'>\The [user] removes \the [stage_name] from \the [src]!</span>", "<span class='notice'>You remove \the [stage_name] from \the [src]</span>")
+						removing.loc = get_turf(src.loc)
```
`get_turf(src.loc)` is the same as `get_turf(src)`.

Techhead0 commented on code/modules/assembly/holder.dm:
```diff
@@ -1,277 +1,850 @@
+						stage = 0
+						stage_name = ""
+						removing = null
+						acting = 0
```
Most of this if statement is the same as the else, except for the order. The only difference seems to be where `removing` ends up.

Techhead0 commented on code/modules/assembly/igniter.dm:
```diff
@@ -2,36 +2,108 @@
+		O.loc = src
+		qdel(O)
+		stage = 3
+		return
```
Putting the stages in order would help with readability.

Techhead0 commented on code/modules/assembly/igniter.dm:
```diff
@@ -2,36 +2,108 @@
+		user << "<span class='notice'>You successfully install an explosive mechanism into \the [src]!</span>"
+		user.drop_item()
+		O.loc = src
+		qdel(O)
```
These three lines (and similar sets of lines) need to be looked at. Why bother changing the `loc` if you're going to immediately qdel the item right after?

Inforsaken replied:
My computer suffers from the terminal illness of slowness, and so garbage collection is sometimes slow and causes objects to remain a fair time after deletion. So it's essentially hiding the object, which is processed instantly, until garbage collection to get to it. (If it's counter-productive I can remove it, but it might help if the server lags or somethin', idk. Helps me aesthetically too.)

Techhead0 commented on code/modules/assembly/logic_circuits.dm:
```diff
@@ -0,0 +1,634 @@
+					if("!=")
+						if(!(value1 != value4 && value2 != value4) && (value1 != value4 || value2 != value4))
+							send_data(list(value3))
+							received = 1
```
The nested switches here are a lot of lines and repeated code.
```
var/op2_1 = 0
var/op2_2 = 0
switch(operation2)
	if("=")
		op2_1 = (value1 == value4)
		op2_2 = (value2 == value4)
...
switch(operation)
	if("AND")
		if(op2_1 && op2_2)
			send_data(list(value3))
			recieved = 1
...
```
Also, if comparing two boolean values, `(!(a&&b) && (a||b))` is equivalent to `(a != b)`.

Inforsaken replied:
I need to re-do logic circuits entirely, they're horrendously done. But honestly, they were such a pain to do in the first place, kinda hard to find the motivation to re-do them.

fludd12 commented:
Not to generate more work for you, but I tried this out on a local, and here are a few problems (which you may already know about!)

- You can't seem to build any of the assembly holders
- You can't modify assembly holders or parts without viewvariables, since the .tmpl files weren't included
- The radio seemed to give a string of unparsed HTML in common, rather than what it input. Might've just been because I was trying to make it work with a piece of paper in a scanner, though.

Immense respect for coding all this, though. The system seems like it'll work great with just a few tweaks (and a UI, of course).

Inforsaken commented:
Ahah, forgot to include the UI when I updated to the latest build sorry about that. Also shouldn't be any more HTML in the radio or speaker. Kinda have no idea how recipes work, so I was hoping someone would point that out with a way to do it.. Hehe

Esoudiere committed:
-Fixed various errors, courtesy of TechHead &

GinjaNinja32 commented:
![](http://nyx.gn32.uk/i/2016-02-19_15:55:40.png)

Any PR that hits this without being an automatic-global-change PR is too big. I'm 99% sure you do not need this much code for the features you're adding.

Techhead0 commented:
So, stuff I commented on, like usr/user, absolute/relative pathing and loc=/forcemove(), needed to be fixed in more than just the one spot I mentioned them. I wasn't going to leave a comment in all of the 40 places you used `loc =`, for example.

Inforsaken commented:
Well its something like 30 different objects being added, the majority of them being over 100 lines, plus the intelligun which is another ~1500. Still, I do not think it should be as big as it is either and I did state in the PR message that there would be alot of problems with it.

I did replace all the `.loc =` with `.forcemove(get_turf(src))` except in instances where an object is being placed inside of another's contents list. As for 'usr/user' I kinda assumed it was a lone case, sorry. I looked through and replaced the other 2 instances. Relative & absolute pathing are not really compatible with eachother, so where relative has been used it often means I've copied something from another file with relative pathing. I'll go ahead and fix it though, as annoying a task as it is.

comma commented:
I think most of diff comes from moving all the assembly files into /new dir, changed bits or not, and all being done in one big commit. Solution would be to travel back in time and remind yourself to do atomic commits, but that might be hard to accomplish with current tech. Maybe soft-resetting to beginning and commiting changes piece by piece, if they are atomic enough.

Inforsaken commented:
Splitting into seperate pull requests for easy reading.
